File: src/extraction/schemas.ts
import { z } from "zod";

export const BoundingBoxSchema = z.object({ x: z.number(), y: z.number(), width: z.number(), height: z.number() });

export const ExtractedValueSchema = z.object({
  rawValue: z.string().nullable(),
  normalisedValue: z.string().nullable().optional(),
  minutes: z.number().int().nullable().optional(),
  confidence: z.number().min(0).max(1).nullable(),
  boundingBox: BoundingBoxSchema.nullable().optional(),
  alternativeInterpretations: z.array(z.string()).optional()
});

export const DailySheetRowSchema = z.object({
  rowIndex: z.number().int(),
  rawEmployeeName: z.string().nullable(),
  matchedEmployeeId: z.string().nullable().optional(),
  matchedEmployeeName: z.string().nullable().optional(),
  employeeMatchConfidence: z.number().min(0).max(1).nullable().optional(),
  start: ExtractedValueSchema.nullable(),
  finish: ExtractedValueSchema.nullable(),
  break: ExtractedValueSchema.nullable(),
  reviewRequired: z.boolean(),
  reviewReasons: z.array(z.string())
});

export const DailySheetExtractionSchema = z.object({
  documentType: z.literal("daily_sheet"),
  date: z.string().nullable(),
  providerName: z.string(),
  processedImagePath: z.string().nullable().optional(),
  qualityWarnings: z.array(z.string()),
  detectedText: z.array(z.object({ text: z.string(), confidence: z.number().nullable(), boundingBox: BoundingBoxSchema.nullable() })).default([]),
  rows: z.array(DailySheetRowSchema),
  unresolvedFields: z.array(z.string()).default([])
});

export const RosterExtractionSchema = z.object({
  documentType: z.literal("roster"),
  providerName: z.string(),
  qualityWarnings: z.array(z.string()),
  rows: z.array(DailySheetRowSchema)
});

export const PayslipExtractionSchema = z.object({
  documentType: z.literal("payslip"),
  providerName: z.string(),
  employeeName: z.string().nullable(),
  ordinaryPaidHours: ExtractedValueSchema.nullable(),
  sundayPaidHours: ExtractedValueSchema.nullable(),
  otherPaidHours: ExtractedValueSchema.nullable(),
  totalPaidHours: ExtractedValueSchema.nullable(),
  rawLabels: z.array(z.string()).default([]),
  reviewRequired: z.boolean(),
  reviewReasons: z.array(z.string())
});

export type DailySheetExtraction = z.infer<typeof DailySheetExtractionSchema>;
export type RosterExtraction = z.infer<typeof RosterExtractionSchema>;
export type PayslipExtraction = z.infer<typeof PayslipExtractionSchema>;

File: ocr-service/app/schemas.py
from pydantic import BaseModel, Field
from typing import Literal


class BoundingBox(BaseModel):
    x: float
    y: float
    width: float
    height: float


class TextDetection(BaseModel):
    text: str
    confidence: float | None = Field(default=None, ge=0, le=1)
    boundingBox: BoundingBox | None = None


class ExtractedValue(BaseModel):
    rawValue: str | None
    normalisedValue: str | None = None
    minutes: int | None = None
    confidence: float | None = Field(default=None, ge=0, le=1)
    boundingBox: BoundingBox | None = None


class DailyRow(BaseModel):
    rowIndex: int
    rawEmployeeName: str | None
    matchedEmployeeId: str | None = None
    matchedEmployeeName: str | None = None
    employeeMatchConfidence: float | None = None
    start: ExtractedValue | None
    finish: ExtractedValue | None
    break_: ExtractedValue | None = Field(default=None, alias="break")
    reviewRequired: bool
    reviewReasons: list[str]


class ExtractionRequest(BaseModel):
    filePath: str
    mimeType: str
    expectedDate: str | None = None
    knownEmployees: list[dict] = []
    expectedColumnNames: list[str] = []
    rosterContext: dict | None = None


class DailySheetResponse(BaseModel):
    documentType: Literal["daily_sheet"] = "daily_sheet"
    date: str | None = None
    providerName: str = "paddle-ocr"
    processedImagePath: str | None = None
    qualityWarnings: list[str]
    detectedText: list[TextDetection]
    rows: list[DailyRow]
    unresolvedFields: list[str]


class RosterResponse(BaseModel):
    documentType: Literal["roster"] = "roster"
    providerName: str = "paddle-ocr"
    qualityWarnings: list[str]
    rows: list[DailyRow]


class PayslipResponse(BaseModel):
    documentType: Literal["payslip"] = "payslip"
    providerName: str = "paddle-ocr"
    employeeName: str | None = None
    ordinaryPaidHours: ExtractedValue | None = None
    sundayPaidHours: ExtractedValue | None = None
    otherPaidHours: ExtractedValue | None = None
    totalPaidHours: ExtractedValue | None = None
    rawLabels: list[str]
    reviewRequired: bool
    reviewReasons: list[str]


class PreprocessResponse(BaseModel):
    width: int
    height: int
    rotated: bool
    pageDetected: bool
    qualityWarnings: list[str]
